actors/builtin/miner/miner_test.go

```diff
@@ -2052,15 +2052,129 @@ func TestWithdrawBalance(t *testing.T) {
 
 func TestChangePeerID(t *testing.T) {
 	periodOffset := abi.ChainEpoch(100)
+	actor := newHarness(t, periodOffset)
+	builder := builderForHarness(actor).
+		WithBalance(bigBalance, big.Zero())
 
 	t.Run("successfully change peer id", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		newPID := tutil.MakePID("test-change-peer-id")
+		actor.changePeerID(rt, newPID)
+	})
+}
+
+func TestCompactPartitions(t *testing.T) {
+	periodOffset := abi.ChainEpoch(100)
+	actor := newHarness(t, periodOffset)
+	builder := builderForHarness(actor).
+		WithBalance(bigBalance, big.Zero())
+
+	t.Run("successfully compact partitions", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		rt.SetEpoch(200)
+		info := actor.commitAndProveSectors(rt, 2, defaultSectorExpiration, [][]abi.DealID{{10}, {20}})
+
+		st := getState(rt)
+		dlIdx, partIdx1, err := st.FindSector(rt.AdtStore(), info[0].SectorNumber)
+		require.NoError(t, err)
+		_, partIdx2, err := st.FindSector(rt.AdtStore(), info[1].SectorNumber)
+		require.NoError(t, err)
+
+		// add partitions to state
+		partitions := bitfield.NewFromSet([]uint64{partIdx1, partIdx2})
+
+		actor.compactPartitions(rt, dlIdx, partitions)
+	})
+
+	t.Run("fails if deadline is equal to WPoStPeriodDeadlines", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		rt.ExpectAbort(exitcode.ErrIllegalArgument, func() {
+			actor.compactPartitions(rt, miner.WPoStPeriodDeadlines, bitfield.New())
+		})
+	})
+
+	t.Run("fails if deadline is not mutable", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		epoch := abi.ChainEpoch(200)
+		rt.SetEpoch(epoch)
+		rt.ExpectAbort(exitcode.ErrForbidden, func() {
+			actor.compactPartitions(rt, 1, bitfield.New())
+		})
+	})
+
+	t.Run("fails if partition count is above limit", func(t *testing.T) {
+		rt := builder.Build(t)
+		actor.constructAndVerify(rt)
+
+		// partition limit is 4 for the default construction
```

Comment:
Where is this set? It should be 200.

Reply:
Hm. I see. Because we might need to load a bunch of sector infos. That's kind of annoying. We could actually just enforce this as we go along.

Reply:
Actually, we can't. I thought we could count live sectors only, but we need to address dead sectors as well. The current logic is correct, if a bit unfortunate.